common/src/stack/pylib/stack/probepal/main.py

```diff
@@ -0,0 +1,60 @@
+import sys
+from operator import attrgetter
+import importlib
+import pkgutil
+
+from common import UnrecognizedPallet
+
+# general:
+# this package allows for fingerprinting a directory to determine if it contains a stacki pallet
+# the design allows for the dynamic discovery of additional finderprinting probes,
+# along with a means of weighting which probes are used first as well as catching partial matches
+# probes can be added by creating a new python module whose name begins with `prober_` and which
+# has a subclass of the Probe class.  See common.py for more detail. 
+
+
+if __name__ == '__main__':
+	# find and import all 'prober' modules in the directory
+	plugins = {
+		name: importlib.import_module(name)
+		for finder, name, ispkg
+		in pkgutil.iter_modules()
+		if name.startswith('prober')
+	}
+
+	probes = []
+	# within those modules, find and instantiate any PalletProber* classes
+	for module in plugins.values():
+		probes.extend(getattr(module, cls)() for cls in dir(module) if cls.startswith('PalletProber'))
```

Comment:
Could you use `issubclass` here to check if a class is a subclass of `Prober` instead of doing a compare on the class name?

Reply:
```
>>> for module in plugins.values():
...                 probes.extend(getattr(module, cls)() for cls in dir(module) if issubclass(cls, Probe))
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 2, in <genexpr>
TypeError: issubclass() arg 1 must be a class
```

:(

I'll look into this a little further.

Reply:
Maybe something from the `inspect` library can help? https://docs.python.org/3/library/inspect.html

Reply:
Is this an improvement?

```
        probes = []
        for module in plugins.values():
                # Get the listing of all classes defined in the module,
                # and instantiate the ones that are subclasses of Prober
                probes.extend(cls()
                        for _, cls in inspect.getmembers(module, inspect.isclass)
                        if cls.__module__ == module.__name__ and issubclass(cls, Prober)
                )
```

Using `inspect`.  `inspect.getmembers(obj, predicate)` returns all members of `obj` which match `predicate`.

Or, same but without comprehension:

```
        probes = []
        for module in plugins.values():
                # Get the listing of all classes defined in the module,
                # and instantiate the ones that are subclasses of Prober
                for _, cls in inspect.getmembers(module, inspect.isclass):
                        if cls.__module__ == module.__name__ and issubclass(cls, Prober):
                                probes.append(cls())
```

Reply:
It definitely seems to be the safer way to ensure that the class being instantiated subclasses the base which, in theory, provides a stronger guarantee that the methods we expect will be there.

What's the `cls.__module__ == module.__name__` for? Can we somehow get classes that aren't part of the module, as in if we import classes from other modules directly via the `from foo import bar` pattern?

Reply:
The one without the comprehension seems nice and clean.

I would think it would be good enough just to check if the class is a subclass of `Prober` unless somehow you got a class loaded in a module that wasn't part of the module, and also inherited from `Prober`.

Reply:
Yeah, it's needed:

```
>>> inspect.getmembers(module, inspect.isclass)
[('PalletInfo', <class 'common.PalletInfo'>), ('Prober', <class 'common.Prober'>), ('TreeinfoProber', <class 'prober_treeinfo.TreeinfoProber'>)]
>>> inspect.getmembers(module, inspect.isclass)[1][1]
<class 'common.Prober'>
>>> issubclass(Prober, inspect.getmembers(module, inspect.isclass)[1][1])
True
```

Also, base classes seem to be subclasses of themselves....

Reply:
Could potentially use [inspect.getmro()](https://docs.python.org/3/library/inspect.html#inspect.getmro) doing something like `cls in inspect.getmro()[1:]` to exclude the base class. But I don't know if that ends up instantiating duplicates probers if you have something like:
```
a.py
    from stack.probepal.common import Prober
    class MyProber(Prober):
        pass

b.py
    from a import MyProber
```